File: fxgl/src/test/java/com/almasb/fxgl/ecs/EntityTest.java
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.ecs;

import com.almasb.fxgl.ecs.component.BooleanComponent;
import com.almasb.fxgl.ecs.component.DoubleComponent;
import com.almasb.fxgl.ecs.component.Required;
import com.almasb.fxgl.ecs.serialization.SerializableComponent;
import com.almasb.fxgl.ecs.serialization.SerializableControl;
import com.almasb.fxgl.io.serialization.Bundle;
import org.junit.Before;
import org.junit.Test;

import java.util.Optional;

import static org.hamcrest.CoreMatchers.*;
import static org.junit.Assert.*;

/**
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
public class EntityTest {

    private Entity entity;

    @Before
    public void setUp() {
        entity = new Entity();
    }

    @Test
    public void testControls() {
        TestControl control = new TestControl();
        entity.addControl(control);

        Optional<TestControl> maybe = entity.getControl(TestControl.class);
        assertTrue(maybe.isPresent());
        assertEquals(control, maybe.get());

        boolean result = entity.removeControl(TestControl.class);
        assertFalse(entity.getControl(TestControl.class).isPresent());
        assertTrue(result);

        entity.addControl(control);
        maybe = entity.getControl(TestControl.class);
        assertTrue(maybe.isPresent());
        assertEquals(control, maybe.get());

        entity.removeAllControls();
        assertFalse(entity.getControl(TestControl.class).isPresent());

        result = entity.removeControl(TestControl.class);
        assertFalse(result);
    }

    @Test
    public void testGetControl() {
        TestControl control = new TestControl();
        entity.addControl(control);

        assertEquals(control, entity.getControl(TestControl.class).get());
        assertEquals(control, entity.getControlUnsafe(TestControl.class));
    }

    @Test
    public void testGetControls() {
        TestControl control = new TestControl();
        ControlAddingControl control2 = new ControlAddingControl();
        ControlRemovingControl control3 = new ControlRemovingControl();

        entity.addControl(control);
        entity.addControl(control2);
        entity.addControl(control3);

        assertThat(entity.getControls(), hasItems(control, control2, control3));
    }

    @Test(expected = IllegalArgumentException.class)
    public void testControlsAnonymous() {
        entity.addControl(new Control() {
            @Override
            public void onAdded(Entity entity) {}

            @Override
            public void onUpdate(Entity entity, double tpf) {}

            @Override
            public void onRemoved(Entity entity) {}
        });
    }

    @Test(expected = IllegalArgumentException.class)
    public void testControlsDuplicate() {
        entity.addControl(new TestControl());
        entity.addControl(new TestControl());
    }

    @Test
    public void testComponents() {
        HPComponent hp = new HPComponent(100);
        entity.addComponent(hp);

        Optional<HPComponent> maybe = entity.getComponent(HPComponent.class);
        assertTrue(maybe.isPresent());
        assertEquals(hp, maybe.get());

        boolean result = entity.removeComponent(HPComponent.class);
        assertTrue(result);
        assertFalse(entity.getComponent(HPComponent.class).isPresent());

        entity.addComponent(hp);
        maybe = entity.getComponent(HPComponent.class);
        assertTrue(maybe.isPresent());
        assertEquals(hp, maybe.get());

        entity.removeAllComponents();
        assertFalse(entity.getComponent(HPComponent.class).isPresent());

        result = entity.removeComponent(HPComponent.class);
        assertFalse(result);
    }

    @Test
    public void testGetComponent() {
        TestComponent component = new TestComponent();
        entity.addComponent(component);

        assertEquals(component, entity.getComponent(TestComponent.class).get());
        assertEquals(component, entity.getComponentUnsafe(TestComponent.class));
    }

    @Test(expected = IllegalStateException.class)
    public void testRequiredAllMissing() {
        entity.addComponent(new ArmorComponent());
    }

    @Test(expected = IllegalStateException.class)
    public void testRequiredOneMissing() {
        entity.addComponent(new TestComponent());
        entity.addComponent(new ArmorComponent());
    }

    @Test
    public void testRequired() {
        entity.addComponent(new TestComponent());
        entity.addComponent(new HPComponent(33));
        entity.addComponent(new ArmorComponent());

        assertTrue(true);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testRequiredRemovedComponent() {
        entity.addComponent(new TestComponent());
        entity.addComponent(new HPComponent(33));
        entity.addComponent(new ArmorComponent());

        entity.removeComponent(HPComponent.class);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testRequiredRemovedControl() {
        entity.addComponent(new HPComponent(33));
        entity.addControl(new HPControl());

        entity.removeComponent(HPComponent.class);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testComponentsAnonymous() {
        entity.addComponent(new Component() {
            @Override
            public void onAdded(Entity entity) {}

            @Override
            public void onRemoved(Entity entity) {}
        });
    }

    @Test(expected = IllegalArgumentException.class)
    public void testComponentsDuplicate() {
        HPComponent hp1 = new HPComponent(100);
        HPComponent hp2 = new HPComponent(135);

        entity.addComponent(hp1);
        entity.addComponent(hp2);
    }

    @Test
    public void testRemoveAllComponents() {
        entity.addComponent(new TestComponent());
        entity.addComponent(new HPComponent(33));

        assertTrue(entity.hasComponent(TestComponent.class));
        assertTrue(entity.hasComponent(HPComponent.class));

        entity.removeAllComponents();

        assertFalse(entity.hasComponent(TestComponent.class));
        assertFalse(entity.hasComponent(HPComponent.class));
    }

    @Test
    public void testRemoveAllControls() {
        entity.addComponent(new HPComponent(33));

        entity.addControl(new TestControl());
        entity.addControl(new HPControl());

        assertTrue(entity.hasControl(TestControl.class));
        assertTrue(entity.hasControl(HPControl.class));

        entity.removeAllControls();

        assertFalse(entity.hasControl(TestControl.class));
        assertFalse(entity.hasControl(HPControl.class));
    }

    @Test
    public void testConcurrentModificationControl() {
        entity.addControl(new ControlAddingControl());
        entity.update(0.017);
    }

    @Test(expected = IllegalStateException.class)
    public void testConcurrentModificationControl2() {
        entity.addControl(new TestControl());
        entity.addControl(new ControlRemovingControl());
        entity.update(0.017);
    }

    @Test
    public void testSetControlsEnabled() {
        entity.addControl(new ControlAddingControl());
        entity.setControlsEnabled(false);
        entity.update(0.017);
    }

    @Test
    public void testControlPause() {
        Control c = new ControlAddingControl();
        entity.addControl(c);

        c.pause();
        entity.update(0.017);
    }

    @Test
    public void testControlResume() {
        Control c = new ControlAddingControl();
        entity.addControl(c);

        c.pause();
        c.resume();
        entity.update(0.017);
    }

    @Test
    public void testComponentListener() {
        HPComponent hp = new HPComponent(20);

        ModuleListener listener = new ModuleListener() {
            @Override
            public void onAdded(Component component) {
                assertEquals(HPComponent.class, component.getClass());

                ((HPComponent)component).setValue(10);
            }

            @Override
            public void onRemoved(Component component) {
                assertEquals(HPComponent.class, component.getClass());

                ((HPComponent)component).setValue(0);
            }
        };

        entity.addModuleListener(listener);

        entity.addComponent(hp);
        assertThat(hp.getValue(), is(10.0));

        entity.removeComponent(HPComponent.class);
        assertThat(hp.getValue(), is(0.0));

        entity.removeModuleListener(listener);

        entity.addModuleListener(listener);
        assertThat(hp.getValue(), is(0.0));
    }

    @Test
    public void testControlListener() {
        HPControl control = new HPControl();

        ModuleListener listener = new ModuleListener() {
            @Override
            public void onAdded(Control control) {
                assertEquals(HPControl.class, control.getClass());

                ((HPControl)control).value = 10;
            }

            @Override
            public void onRemoved(Control control) {
                assertEquals(HPControl.class, control.getClass());

                ((HPControl)control).value = 20;
            }
        };

        entity.addModuleListener(listener);
        entity.addComponent(new HPComponent(33));

        entity.addControl(control);
        assertEquals(10, control.value, 0);

        entity.removeControl(HPControl.class);
        assertEquals(20, control.value, 0);

        entity.removeModuleListener(listener);

        entity.addControl(control);
        assertEquals(20, control.value, 0);
    }

    @Test
    public void testSave() {
        entity.addComponent(new GravityComponent(true));
        entity.addComponent(new CustomDataComponent("SerializationData"));
        entity.addControl(new CustomDataControl("SerializableControl"));

        Bundle bundle = new Bundle("Entity01234");
        entity.save(bundle);

        Entity entity2 = new Entity();
        entity2.addComponent(new GravityComponent(false));
        entity2.addComponent(new CustomDataComponent(""));
        entity2.addControl(new CustomDataControl(""));

        entity2.load(bundle);

        assertThat(entity2.getControlUnsafe(CustomDataControl.class).data, is("SerializableControl"));
        assertThat(entity2.getComponentUnsafe(CustomDataComponent.class).data, is("SerializationData"));
        assertThat(entity2.getComponentUnsafe(GravityComponent.class).getValue(), is(true));
    }

    @Test
    public void testSaveNoSuchComponent() {
        Bundle bundle = new Bundle("Entity01234");
        entity.save(bundle);

        Entity entity2 = new Entity();
        entity2.addComponent(new CustomDataComponent(""));

        // even if bundle has no such component coz entity1 did not have it
        // we do not fail, but log a warning
        entity2.load(bundle);
    }

    @Test
    public void testProperties() {
        entity.setProperty("hp", 30);
        assertThat(entity.getProperty("hp"), is(30));

        entity.setProperty("hp", true);
        assertThat(entity.getProperty("hp"), is(true));
    }

    @Test
    public void testActiveCallbacks() {
        HPControl hp = new HPControl();

        assertFalse(entity.activeProperty().get());

        entity.setOnActive(() -> hp.value = 30.0);
        assertThat(hp.value, is(0.0));

        GameWorld world = new GameWorld();
        world.addEntity(entity);
        assertThat(hp.value, is(30.0));
        assertTrue(entity.activeProperty().get());

        entity.setOnNotActive(() -> hp.value = -50.0);
        assertThat(hp.value, is(30.0));

        world.removeEntity(entity);
        assertThat(hp.value, is(-50.0));
        assertFalse(entity.activeProperty().get());

        entity.setOnNotActive(() -> hp.value = -33.0);
        assertThat(hp.value, is(-33.0));

        world.addEntity(entity);
        entity.setOnActive(() -> hp.value = 99.0);
        assertThat(hp.value, is(99.0));
    }

    @Test(expected = IllegalArgumentException.class)
    public void testNoPropertyKey() {
        entity.getProperty("no_key");
    }

    @Test
    public void testIntegrity() {
        int count = 0;

        GameWorld world = new GameWorld();

        world.addEntity(entity);
        world.removeEntity(entity);

        try {
            entity.addComponent(new HPComponent(23));
        } catch (IllegalStateException e) {
            count++;
        }

        assertThat(count, is(1));

        try {
            entity.removeFromWorld();
        } catch (IllegalStateException e) {
            count++;
        }

        assertThat(count, is(2));
    }

    @Test
    public void testRemoveFromWorld() {
        GameWorld world = new GameWorld();

        world.addEntity(entity);
        assertThat(world.getEntities(), hasItems(entity));

        entity.removeFromWorld();
        assertThat(world.getEntities(), not(hasItems(entity)));

        Entity ee = new Entity();
        ee.addControl(new EntityRemovingControl());

        world.addEntity(ee);
        world.onUpdate(0);

        assertThat(world.getEntities(), not(hasItems(ee)));
    }

    @Test
    public void testCopy() {
        entity.addComponent(new HPComponent(33));

        Entity e2 = entity.copy();

        assertThat(e2.hasComponent(HPComponent.class), is(true));
        assertThat(e2.getComponentUnsafe(HPComponent.class).getValue(), is(33.0));
    }

    @Test
    public void testToString() {
        HPComponent component = new HPComponent(33);
        HPControl control = new HPControl();

        entity.addComponent(component);
        entity.addControl(control);

        String toString = entity.toString();

        assertThat(toString, containsString(component.toString()));
        assertThat(toString, containsString(control.toString()));
    }

    private class TestControl extends Control {
        @Override
        public void onUpdate(Entity entity, double tpf) {}
    }

    private class ControlAddingControl extends Control {
        @Override
        public void onUpdate(Entity entity, double tpf) {
            entity.addControl(new TestControl());
        }
    }

    private class ControlRemovingControl extends Control {
        @Override
        public void onUpdate(Entity entity, double tpf) {
            entity.removeAllControls();
        }
    }

    private class EntityRemovingControl extends Control {
        @Override
        public void onUpdate(Entity entity, double tpf) {
            entity.removeFromWorld();
        }
    }

    @Required(HPComponent.class)
    private class HPControl extends Control {
        private double value = 0;

        @Override
        public void onUpdate(Entity entity, double tpf) {}
    }

    private class TestComponent extends Component {
    }

    private class HPComponent extends DoubleComponent implements CopyableComponent<HPComponent> {
        HPComponent(double value) {
            super(value);
        }

        @Override
        public HPComponent copy() {
            return new HPComponent(getValue());
        }
    }

    @Required(TestComponent.class)
    @Required(HPComponent.class)
    private class ArmorComponent extends Component {
    }

    private class GravityComponent extends BooleanComponent {
        GravityComponent(boolean value) {
            super(value);
        }
    }

    public static class CustomDataComponent extends Component implements SerializableComponent {

        public String data;

        CustomDataComponent(String data) {
            this.data = data;
        }

        @Override
        public void write(Bundle bundle) {
            bundle.put("data", data);
        }

        @Override
        public void read(Bundle bundle) {
            data = bundle.get("data");
        }
    }

    public static class CustomDataControl extends Control implements SerializableControl {

        public String data;

        CustomDataControl(String data) {
            this.data = data;
        }

        @Override
        public void onUpdate(Entity entity, double tpf) {

        }

        @Override
        public void write(Bundle bundle) {
            bundle.put("data", data);
        }

        @Override
        public void read(Bundle bundle) {
            data = bundle.get("data");
        }
    }
}

File: fxgl/src/test/kotlin/com/almasb/fxgl/entity/EntityTest.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.entity

import com.almasb.fxgl.entity.component.BooleanComponent
import com.almasb.fxgl.entity.component.DoubleComponent
import com.almasb.fxgl.entity.component.Required
import com.almasb.fxgl.entity.serialization.SerializableComponent
import com.almasb.fxgl.entity.serialization.SerializableControl
import com.almasb.fxgl.io.serialization.Bundle
import org.hamcrest.CoreMatchers.*
import org.hamcrest.MatcherAssert.assertThat
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

/**
 *
 *
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
class EntityTest {

    private lateinit var entity: Entity

    @BeforeEach
    fun setUp() {
        entity = Entity()
    }

    @Test
    fun `Add component`() {
        val comp = TestComponent()
        entity.addComponent(comp)

        assertThat(entity.components, hasItem(comp))
    }

    @Test
    fun `Add component fails if same type already exists`() {
        val exception = assertThrows(IllegalArgumentException::class.java) {
            entity.addComponent(TestComponent())
            entity.addComponent(TestComponent())
        }
    }

    @Test
    fun `Add component fails if component is anonymous`() {
        assertThrows(IllegalArgumentException::class.java, {
            entity.addComponent(object : Component() {})
        })
    }

    @Test
    fun `Add component fails if required component is missing`() {
        assertThrows(IllegalStateException::class.java, {
            entity.addComponent(RequireTestComponent())
        })
    }

    @Test
    fun `Get component`() {
        val comp = TestComponent()
        entity.addComponent(comp)

        assertThat(entity.getComponent(TestComponent::class.java), `is`(comp))
    }

    @Test
    fun `Get component optional`() {
        val comp = TestComponent()
        entity.addComponent(comp)

        assertThat(entity.getComponentOptional(TestComponent::class.java).get(), `is`(comp))
    }

    @Test
    fun `Has component`() {
        entity.addComponent(TestComponent())

        assertTrue(entity.hasComponent(TestComponent::class.java))
    }

    @Test
    fun `Remove component returns true if removed`() {
        val comp = TestComponent()
        entity.addComponent(comp)

        val isRemoved = entity.removeComponent(TestComponent::class.java)

        assertTrue(isRemoved)
        assertThat(entity.components, not(hasItem(comp)))
    }

    @Test
    fun `Remove component returns false if not found`() {
        val isRemoved = entity.removeComponent(TestComponent::class.java)

        assertFalse(isRemoved)
    }

    @Test
    fun `Remove component fails if component is required by another component`() {
        entity.addComponent(TestComponent())
        entity.addComponent(RequireTestComponent())

        assertThrows(IllegalArgumentException::class.java, {
            entity.removeComponent(TestComponent::class.java)
        })
    }

    @Test
    fun `Remove component fails if component is required by a control`() {
        entity.addComponent(HPComponent(0.0))
        entity.addControl(HPControl())

        assertThrows(IllegalArgumentException::class.java, {
            entity.removeComponent(HPComponent::class.java)
        })
    }

//    @Test
//    fun `Remove all components`() {
//        val comp1 = TestComponent()
//        entity.addComponent(comp1)
//
//        val comp2 = RequireTestComponent()
//        entity.addComponent(comp2)
//
//        entity.removeAllComponents()
//
//        assertThat(entity.components, not(hasItems(comp1, comp2)))
//    }

    @Test
    fun `Add control`() {
        val control = TestControl()
        entity.addControl(control)

        assertThat(entity.controls, hasItem(control))
    }

    @Test
    fun `Add control fails if same type exists`() {
        val control = TestControl()
        entity.addControl(control)

        assertThrows(IllegalArgumentException::class.java, {
            entity.addControl(control)
        })
    }

    @Test
    fun `Add control fails if control is anonymous`() {
        assertThrows(IllegalArgumentException::class.java, {
            entity.addControl(object : Control() {
                override fun onUpdate(entity: Entity, tpf: Double) { }
            })
        })
    }

    @Test
    fun `Add control fails if required module is missing`() {
        assertThrows(IllegalStateException::class.java, {
            entity.addControl(HPControl())
        })
    }

    @Test
    fun `Add control with required module`() {
        entity.addComponent(HPComponent(0.0))
        entity.addControl(HPControl())

        assertTrue(entity.hasControl(HPControl::class.java))
    }

    @Test
    fun `Add control fails if within update of another control`() {
        val control = ControlAddingControl()
        entity.addControl(control)

        assertThrows(IllegalStateException::class.java, {
            entity.update(0.0)
        })
    }

    @Test
    fun `Remove control returns true if removed`() {
        val control = TestControl()
        entity.addControl(control)
        val isRemoved = entity.removeControl(TestControl::class.java)

        assertTrue(isRemoved)
        assertThat(entity.controls, not(hasItem(control)))
    }

    @Test
    fun `Remove control returns false if control not found`() {
        val isRemoved = entity.removeControl(TestControl::class.java)

        assertFalse(isRemoved)
    }

    @Test
    fun `Remove control fails if within update of another control`() {
        val control = ControlRemovingControl()
        entity.addControl(control)

        assertThrows(IllegalStateException::class.java, {
            entity.update(0.0)
        })
    }

    @Test
    fun `Has control`() {
        val control = TestControl()
        entity.addControl(control)

        assertTrue(entity.hasControl(TestControl::class.java))
    }

    @Test
    fun `Get control`() {
        val control = TestControl()
        entity.addControl(control)

        assertThat(entity.getControl(TestControl::class.java), `is`(control))
    }

    @Test
    fun `Get control optional`() {
        val control = TestControl()
        entity.addControl(control)

        assertThat(entity.getControlOptional(TestControl::class.java).get(), `is`(control))
    }

//    @Test
//    fun `Remove all controls`() {
//        val control = TestControl()
//        entity.addControl(control)
//
//        val control2 = ControlAddingControl()
//        entity.addControl(control2)
//
//        entity.removeAllControls()
//
//        assertThat(entity.controls, not(hasItems(control, control2)))
//    }

//    @Test
//    fun `Remove all controls fails if within update of another control`() {
//        entity.addControl(AllControlRemovingControl())
//        entity.update(0.0)
//    }

    @Test
    fun `Update controls`() {
        val control = ValueUpdateControl()
        entity.addControl(control)

        entity.update(0.017)

        assertThat(control.value, `is`(0.017))
    }

    @Test
    fun `Disable controls`() {
        val control = ValueUpdateControl()
        entity.addControl(control)

        entity.setControlsEnabled(false)

        entity.update(0.017)

        assertThat(control.value, `is`(0.0))
    }

    @Test
    fun `Pause control`() {
        val control = ValueUpdateControl()
        entity.addControl(control)

        control.pause()

        entity.update(0.017)

        assertThat(control.value, `is`(0.0))
    }

    @Test
    fun `Resume control`() {
        val control = ValueUpdateControl()
        entity.addControl(control)

        control.pause()
        control.resume()

        entity.update(0.017)

        assertThat(control.value, `is`(0.017))
    }

    @Test
    fun `Get property returns correct value`() {
        entity.setProperty("hp", 30)

        assertThat(entity.getProperty("hp"), `is`(30))
    }

    @Test
    fun `Get property optional returns correct value`() {
        entity.setProperty("hp", 30)

        assertThat(entity.getPropertyOptional<Int>("hp").get(), `is`(30))
    }

    @Test
    fun `Get property returns null if key not found`() {
        assertThat(entity.getProperty<Any>("no_key"), nullValue())
    }

    @Test
    fun `Get property returns null if value is null`() {
        entity.setProperty("lastTime", null)

        assertThat(entity.getProperty<Any>("lastTime"), nullValue())
    }

    @Test
    fun `Get property optional returns empty if key not found`() {
        val value = entity.getPropertyOptional<Any>("no_key")

        assertFalse(value.isPresent)
    }

    @Test
    fun `Module listener is notified on component add`() {
        val hp = HPComponent(20.0)

        val listener = object : ModuleListener {
            override fun onAdded(component: Component) {
                (component as HPComponent).value = 10.0
            }
        }

        entity.addModuleListener(listener)

        entity.addComponent(hp)

        assertThat(hp.value, `is`(10.0))
    }

    @Test
    fun `Module listener is notified on component remove`() {
        val hp = HPComponent(20.0)

        val listener = object : ModuleListener {
            override fun onRemoved(component: Component) {
                (component as HPComponent).value = 0.0
            }
        }

        entity.addModuleListener(listener)

        entity.addComponent(hp)

        entity.removeComponent(HPComponent::class.java)

        assertThat(hp.value, `is`(0.0))
    }

    @Test
    fun `Module listener is notified on control add`() {
        val control = HPControl(0)

        val listener = object : ModuleListener {
            override fun onAdded(control: Control) {
                (control as HPControl).value = 10
            }
        }

        entity.addModuleListener(listener)
        entity.addComponent(HPComponent(33.0))

        entity.addControl(control)

        assertThat(control.value, `is`(10))
    }

    @Test
    fun `Module listener is notified on control remove`() {
        val control = HPControl()

        val listener = object : ModuleListener {
            override fun onRemoved(control: Control) {
                (control as HPControl).value = 20
            }
        }

        entity.addModuleListener(listener)
        entity.addComponent(HPComponent(33.0))

        entity.addControl(control)

        entity.removeControl(HPControl::class.java)

        assertThat(control.value, `is`(20))
    }

    @Test
    fun `Module listener is removed correctly`() {
        val hp = HPComponent(20.0)

        val listener = object : ModuleListener {
            override fun onAdded(component: Component) {
                (component as HPComponent).value = 0.0
            }
        }

        entity.addModuleListener(listener)
        entity.removeModuleListener(listener)

        entity.addComponent(hp)

        assertThat(hp.value, `is`(20.0))
    }

    @Test
    fun `Save and load`() {
        entity.addComponent(GravityComponent(true))
        entity.addComponent(CustomDataComponent("SerializationData"))
        entity.addControl(CustomDataControl("SerializableControl"))

        val bundle = Bundle("Entity01234")
        entity.save(bundle)

        val entity2 = Entity()
        entity2.addComponent(GravityComponent(false))
        entity2.addComponent(CustomDataComponent(""))
        entity2.addControl(CustomDataControl(""))

        entity2.load(bundle)

        assertThat(entity2.getControl(CustomDataControl::class.java).data, `is`("SerializableControl"))
        assertThat(entity2.getComponent(CustomDataComponent::class.java).data, `is`("SerializationData"))
        assertThat(entity2.getComponent(GravityComponent::class.java).value, `is`(true))
    }

    @Test
    fun `Load from entity with different components`() {
        val bundle = Bundle("Entity01234")
        entity.save(bundle)

        val entity2 = Entity()
        entity2.addComponent(CustomDataComponent(""))

        // if bundle has no such component coz entity1 did not have it, we ignore it
        entity2.load(bundle)

        assertThat(entity2.components.size(), `is`(5 + 1))
    }

    @Test
    fun `Copying an entity also copies components and controls`() {
        entity.addComponent(HPComponent(33.0))
        entity.addControl(HPControl(10))

        val e2 = entity.copy()

        assertTrue(e2.hasComponent(HPComponent::class.java))
        assertTrue(e2.hasControl(HPControl::class.java))

        assertThat(e2.getComponent(HPComponent::class.java).value, `is`(33.0))
        assertThat(e2.getControl(HPControl::class.java).value, `is`(10))
    }

    @Test
    fun testToString() {
        val component = HPComponent(33.0)
        val control = HPControl()

        entity.addComponent(component)
        entity.addControl(control)

        val toString = entity.toString()

        assertThat(toString, containsString(component.toString()))
        assertThat(toString, containsString(control.toString()))
    }

    /* INTERACTIONS WITH GAME WORLD */

    @Test
    fun `Entity world is null when created`() {
        assertThat(entity.world, nullValue())
    }

    @Test
    fun `Entity world is set when entity added to game world`() {
        val world = GameWorld()
        world.addEntity(entity)

        assertThat(entity.world, `is`(world))
    }

    @Test
    fun `Entity is not active when created`() {
        assertFalse(entity.isActive)
    }

    @Test
    fun `Entity is active in the same frame when added to game world`() {
        val world = GameWorld()
        world.addEntity(entity)

        assertTrue(entity.isActive)
    }

    @Test
    fun `Entity is not active in the same frame when removed from game world`() {
        val world = GameWorld()
        world.addEntity(entity)
        world.removeEntity(entity)

        assertFalse(entity.isActive)
    }

    @Test
    fun `Set on active callback fires in the same frame when added to world`() {
        var value = 0

        entity.setOnActive { value = 5 }

        val world = GameWorld()
        world.addEntity(entity)

        assertThat(value, `is`(5))
    }

    @Test
    fun `Set on not active callback fires in the same frame when removed from world`() {
        var value = 0

        entity.setOnNotActive { value = 5 }

        val world = GameWorld()
        world.addEntity(entity)
        world.removeEntity(entity)

        assertThat(value, `is`(5))
    }

    @Test
    fun `Set on active callback does not fire if entity already in world`() {
        val world = GameWorld()
        world.addEntity(entity)

        var value = 0

        entity.setOnActive { value = 5 }

        assertThat(value, `is`(0))
    }

    @Test
    fun `Set on not active callback does not fire if entity already not in world`() {
        var value = 0

        entity.setOnNotActive { value = 5 }

        assertThat(value, `is`(0))
    }

    @Test
    fun `Active listener fires in the same frame when added to world`() {
        var value = 0

        entity.activeProperty().addListener { _, _, newValue ->
            if (newValue)
                value = 5
        }

        val world = GameWorld()
        world.addEntity(entity)

        assertThat(value, `is`(5))
    }

    @Test
    fun `Active listener fires in the same frame when removed from world`() {
        var value = 0

        entity.activeProperty().addListener { _, _, newValue ->
            if (!newValue)
                value = 5
        }

        val world = GameWorld()
        world.addEntity(entity)
        world.removeEntity(entity)

        assertThat(value, `is`(5))
    }

    @Test
    fun `Remove from world`() {
        val world = GameWorld()

        world.addEntity(entity)
        entity.removeFromWorld()

        assertThat(world.entities, not(hasItem(entity)))
    }

    @Test
    fun `Remove from world does not fail when entity is not attached to any world`() {
        entity.removeFromWorld()
    }

    @Test
    fun `Components are removed when entity is cleaned`() {
        entity.addComponent(TestComponent())

        entity.clean()

        assertThat(entity.components.size(), `is`(0))
    }

    @Test
    fun `Controls are removed when entity is cleaned`() {
        entity.addControl(TestControl())

        entity.clean()

        assertThat(entity.controls.size(), `is`(0))
    }

    /* MOCK CLASSES */

    private inner class TestControl : Control() {
        override fun onUpdate(entity: Entity, tpf: Double) {}
    }

    private inner class ControlAddingControl : Control() {
        override fun onUpdate(entity: Entity, tpf: Double) {
            entity.addControl(TestControl())
        }
    }

    private inner class ControlRemovingControl : Control() {
        override fun onUpdate(entity: Entity, tpf: Double) {
            entity.removeControl(TestControl::class.java)
        }
    }

//    private inner class AllControlRemovingControl : Control() {
//        override fun onUpdate(entity: Entity, tpf: Double) {
//            entity.removeAllControls()
//        }
//    }

    private inner class EntityRemovingControl : Control() {
        override fun onUpdate(entity: Entity, tpf: Double) {
            entity.removeFromWorld()
        }
    }

    private inner class ValueUpdateControl : Control() {
        var value = 0.0

        override fun onUpdate(entity: Entity, tpf: Double) {
            value += tpf
        }
    }

    @Required(HPComponent::class)
    private inner class HPControl(var value: Int = 0) : Control(), CopyableControl<HPControl> {

        override fun onUpdate(entity: Entity, tpf: Double) {}

        override fun copy(): HPControl {
            return HPControl(value)
        }
    }

    private inner class TestComponent : Component()

    private inner class HPComponent(value: Double) : DoubleComponent(value), CopyableComponent<HPComponent> {

        override fun copy(): HPComponent {
            return HPComponent(value)
        }
    }

    @Required(TestComponent::class)
    //@Required(HPComponent::class)
    private inner class RequireTestComponent : Component()

    private inner class GravityComponent internal constructor(value: Boolean) : BooleanComponent(value)

    class CustomDataComponent constructor(var data: String) : Component(), SerializableComponent {

        override fun write(bundle: Bundle) {
            bundle.put("data", data)
        }

        override fun read(bundle: Bundle) {
            data = bundle.get("data")
        }
    }

    class CustomDataControl constructor(var data: String) : Control(), SerializableControl {

        override fun onUpdate(entity: Entity, tpf: Double) { }

        override fun write(bundle: Bundle) {
            bundle.put("data", data)
        }

        override fun read(bundle: Bundle) {
            data = bundle.get("data")
        }
    }
}